### ISSUE TITLE
Lower F1 Performance After Hyperparameter Tuning (SparseNodeTransformer)

### ISSUE THREAD

**Issue by okwow123** [junhongmit/FraudGT#2]
Hi, thank you for the great work on FraudGT.
I’ve been experimenting with your model using the SparseNodeTransformer architecture on the AML-Small-LI dataset.

I made the following changes from the original configuration:

Increased layers and layers_post_gt from 2 → 3

Adjusted dropout and attn_dropout to slightly lower values

Enabled batch_norm

Increased the loss_fun_weight for the minority class

Used a slightly smaller base_lr (0.0007)

Despite these modifications, my test F1 score decreased compared to the original setting, while the macro-F1 and AUC stayed high (as seen in the attached WandB plots).

Hyperparameter Tuning : https://github.com/okwow123/FraudGT/blob/main/configs/AML-Small-LI/AML-Small-LI-Multi-SparseNodeGT_v2.yaml

<img width="991" height="667" alt="Image" src="https://github.com/user-attachments/assets/777112ed-408c-4eaf-80bc-04f043456635" />
<img width="1000" height="665" alt="Image" src="https://github.com/user-attachments/assets/90158f12-618a-4b44-a00a-3f3761a47398" />

<img width="816" height="506" alt="Image" src="https://github.com/user-attachments/assets/e0716841-6434-4162-8195-5100e8bc9362" />
<img width="829" height="511" alt="Image" src="https://github.com/user-attachments/assets/03c2a554-d90f-47ea-a867-7261029d02d9" />

Do you have any insights on why this might happen?
Could the added model depth or the changed normalization be causing gradient instability or underfitting in some classes?
I’d be truly grateful for any valuable suggestions or ideas you could share to help improve the performance, even if they involve a different approach from what I’ve attempted so far.

Any advice or best practices for tuning this model would be greatly appreciated.
Thanks again for open-sourcing your excellent work.

Best regards,

**Comment by junhongmit** [comment]
Hi okwow123,

Thanks again for your detailed follow-up and for sharing your modifications and WandB logs — it's great to see your deep engagement with the model!

At first glance, one thing that stands out in your config is the setting of `train.neighbor_sizes: [50, 50]`. This configuration means you're only sampling up to **2-hop neighbors** (50 per hop), which may significantly limit the information captured when you use a 3-layer GNN. A 3-layer GNN means to pass messages to up to 3-hop neighbors, but you only have two hops here.

In our experiments, we typically set neighbor_sizes equal to the model depth. For your case, you can try `train.neighbor_sizes: [50, 50, 50]` (i.e., **3-hop sampling**) to allow the model to gather richer context around each node. Reducing the number of hops may lead to underfitting or incomplete local subgraphs, especially for accounts that are part of complex fraud rings, which could explain the drop in test F1 even when AUC and macro-F1 remain relatively high. But increasing the sampled `neighbor_sizes` will also greatly increase the running time, because the number of neighbors will grow exponentially. I would recommend trying all the parameters before moving on to the 3-layer.

A few other suggestions worth considering:

- Model Depth: While deeper architectures can capture more complex patterns, they can also lead to over-smoothing or optimization difficulty, especially in GNNs. If increasing `layers` and `layers_post_gt` from 2 to 3 reduced F1, it’s worth trying 2+2 again while focusing tuning efforts elsewhere. I found in many situations, more `post_gt` layers make things worse, possibly due to overfitting.

- Normalization: Enabling `batch_norm` can help training stability, but may sometimes interfere with attention mechanisms. You could try `layer_norm` or run an ablation to compare both. In my case, I didn't find that it affected the prediction much.

- Minority Class Weighting: It’s good to increase the `loss_fun_weight` for the minority class, but too large a weight can lead to overfitting or poor calibration. A controlled experiment (holding all other parameters the same) with grid search with small increments may help here. I just used the value from the Multi-GNN paper, and found it performs very well already.

- Learning Rate: Your reduced base learning rate (0.0007) is reasonable. You might also try tuning `weight_decay` and adjusting the warm-up steps.

I hope these can help! These are all on top of mind for now. I know the hyperparameter tuning is not fun — I would recommend always performing a controlled experiment, only tune one to see if it makes things better. This kind of "manual gradient descent" I found is usually very effective. A good thing about using WandB is that they have a log history to compare the best F1 and their hyperparameters between runs, so you know which parameter change makes things better.

Best,
Junhong

**Comment by okwow123** [comment]
Your support has been immensely helpful. I sincerely appreciate your response, and I will be sure to share any positive research outcomes with you. I also wish you the best of luck in everything you do.
